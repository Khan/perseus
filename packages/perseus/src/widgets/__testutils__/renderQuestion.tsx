import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
