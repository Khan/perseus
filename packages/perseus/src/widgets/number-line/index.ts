export {default} from "./number-line";
