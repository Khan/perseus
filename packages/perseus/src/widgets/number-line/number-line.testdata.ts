import type {PerseusRenderer} from "../../perseus-types";

export const question1: PerseusRenderer = {
    content:
        "$E=2.5$\n\n**Move the dot to $-E$ on the number line.**\n\n\n[[\u2603 number-line 1]]",
    images: {},
    widgets: {
        "number-line 1": {
            graded: true,
            version: {major: 0, minor: 0},
            static: false,
            type: "number-line",
            options: {
                labelRange: [null, null],
                initialX: null,
                tickStep: 1,
                labelStyle: "decimal",
                labelTicks: true,
                snapDivisions: 2,
                range: [-4, 4],
                static: false,
                correctRel: "eq",
                numDivisions: null,
                divisionRange: [1, 10],
                correctX: -2.5,
            },
            alignment: "default",
        },
    },
};

export const question2: PerseusRenderer = {
    content:
        "$E=2.5$\n\n**Move the dot to $-E$ on the number line.**\n\n\n[[\u2603 number-line 1]]",
    images: {},
    widgets: {
        "number-line 1": {
            type: "number-line",
            alignment: "default",
            static: false,
            graded: true,
            options: {
                static: false,
                range: [0, 1],
                labelRange: [null, null],
                labelStyle: "improper",
                labelTicks: false,
                divisionRange: [1, 12],
                numDivisions: 1,
                snapDivisions: 1,
                tickStep: null,
                correctRel: "eq",
                correctX: 0.5,
                initialX: null,
                showTooltip: false,
                isTickCtrl: true,
            },
        },
    },
};
