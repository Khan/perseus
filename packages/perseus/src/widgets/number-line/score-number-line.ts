import {number as knumber} from "@khanacademy/kmath";

import type {PerseusScore} from "../../types";
import type {
    PerseusNumberLineRubric,
    PerseusNumberLineUserInput,
} from "../../validation.types";

function scoreNumberLine(
    state: PerseusNumberLineUserInput,
    rubric: PerseusNumberLineRubric,
): PerseusScore {
    const range = rubric.range;
    const divisionRange = state.divisionRange;
    const start = rubric.initialX != null ? rubric.initialX : range[0];
    const startRel = rubric.isInequality ? "ge" : "eq";
    const correctRel = rubric.correctRel || "eq";
    const correctPos = knumber.equal(
        state.numLinePosition,
        rubric.correctX || 0,
    );
    const outsideAllowedRange =
        state.numDivisions > divisionRange[1] ||
        state.numDivisions < divisionRange[0];

    // TODO: I don't think isTickCrtl is a thing anymore
    if (state.isTickCrtl && outsideAllowedRange) {
        return {
            type: "invalid",
            message: "Number of divisions is outside the allowed range.",
        };
    }
    if (correctPos && correctRel === state.rel) {
        return {
            type: "points",
            earned: 1,
            total: 1,
            message: null,
        };
    }
    if (state.numLinePosition === start && state.rel === startRel) {
        // We're where we started.
        return {
            type: "invalid",
            message: null,
        };
    }
    return {
        type: "points",
        earned: 0,
        total: 1,
        message: null,
    };
}

export default scoreNumberLine;
