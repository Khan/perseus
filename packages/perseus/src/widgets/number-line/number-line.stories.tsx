import {ServerItemRendererWithDebugUI} from "../../../../../testing/server-item-renderer-with-debug-ui";
import {generateTestPerseusItem} from "../../util/test-utils";

import {question1, question2} from "./number-line.testdata";

import type {Meta, StoryObj} from "@storybook/react";

const meta: Meta = {
    title: "Perseus/Widgets/Number Line",
    component: ServerItemRendererWithDebugUI,
};
export default meta;

type Story = StoryObj<typeof ServerItemRendererWithDebugUI>;

export const BasicQuestion: Story = {
    args: {
        item: generateTestPerseusItem({question: question1}),
    },
};

export const ShowTickController: Story = {
    args: {
        item: generateTestPerseusItem({question: question2}),
    },
};
