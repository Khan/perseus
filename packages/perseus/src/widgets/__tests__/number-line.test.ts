import "@testing-library/jest-dom";

import {testDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import {question1} from "../__testdata__/number-line.testdata";

import {renderQuestion} from "./renderQuestion";

import type {APIOptions} from "../../types";

describe("number-line widget", () => {
    beforeEach(() => {
        // This module complains but doesn't have a real problem
        // with displaying the graphie
        jest.spyOn(console, "log").mockImplementation(() => {});
        jest.spyOn(console, "error").mockImplementation(() => {});

        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("should snapshot", () => {
        // Arrange
        const apiOptions: APIOptions = {
            isMobile: false,
        };

        // Act
        const {container} = renderQuestion(question1, apiOptions);

        // Assert
        expect(container).toMatchSnapshot("first render");
    });

    it("should snapshot on mobile", () => {
        // Arrange
        jest.useRealTimers();

        const apiOptions: APIOptions = {
            isMobile: true,
        };

        // Act
        const {container} = renderQuestion(question1, apiOptions);

        // Assert
        expect(container).toMatchSnapshot("first mobile render");
    });

    it("can be answered correctly", () => {
        // Arrange
        const apiOptions: APIOptions = {
            isMobile: false,
        };
        const {renderer} = renderQuestion(question1, apiOptions);

        // Act
        const [numberLine] = renderer.findWidgets("number-line 1");
        numberLine.movePosition(-2.5);

        // assert
        expect(renderer).toHaveBeenAnsweredCorrectly();
    });

    it("can be answered incorrectly", () => {
        // Arrange
        const apiOptions: APIOptions = {
            isMobile: false,
        };
        const {renderer} = renderQuestion(question1, apiOptions);

        // Act
        const [numberLine] = renderer.findWidgets("number-line 1");
        numberLine.movePosition(3.5);
        // Assert
        expect(renderer).toHaveBeenAnsweredIncorrectly();
    });
});
