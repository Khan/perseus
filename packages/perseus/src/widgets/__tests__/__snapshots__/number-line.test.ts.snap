// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`number-line widget should snapshot on mobile: first mobile render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph perseus-paragraph-centered"
      data-perseus-paragraph-index="0"
    >
      <div
        class="perseus-block-math"
        style="margin-left: -16px; margin-right: -16px;"
      >
        <div
          class="perseus-block-math-inner"
          style="overflow-x: auto; overflow-y: hidden; margin-top: -10px; margin-bottom: -10px; transform: translate3d(0,0,0); padding: 10px 16px 10px 16px;"
        >
          <span
            style="display: block; width: 100%; transform: scale(1, 1); transform-origin: 0 0; opacity: 1;"
          >
            <span
              class="mock-TeX"
            >
              E=2.5
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="1"
    >
      <div
        class="paragraph"
      >
        <strong>
          Move the dot to 
          <span
            style="white-space: nowrap;"
          >
            <span />
            <span
              class="mock-TeX"
            >
              -E
            </span>
            <span />
          </span>
           on the number line.
        </strong>
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="2"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <div
            class="perseus-widget perseus-widget-interactive-number-line"
          >
            <div
              class="graphie-container"
            >
              <div
                class="graphie"
                style="position: relative; width: 288px; height: 80px;"
              >
                <svg
                  height="80"
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); z-index: 2;"
                  version="1.1"
                  width="288"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <desc
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                  >
                    Created with Raphaël
                  </desc>
                  <defs
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                  />
                  <rect
                    fill="#000000"
                    height="80"
                    opacity="0"
                    r="0"
                    rx="0"
                    ry="0"
                    stroke="#000"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); opacity: 0;"
                    width="288"
                    x="0"
                    y="0"
                  />
                </svg>
                <svg
                  height="80"
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                  version="1.1"
                  width="288"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <desc
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                  >
                    Created with Raphaël
                  </desc>
                  <defs
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                  />
                  <path
                    d="M281.05,45.6C281.05,45.6,286.3,40,286.3,40C286.3,40,281.05,34.4,281.05,34.4"
                    fill="none"
                    stroke="#000000"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2; stroke-linejoin: round; stroke-linecap: round;"
                    transform=""
                  />
                  <path
                    d="M144,40C144,40,144,40,286.95,40"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2;"
                  />
                  <path
                    d="M-4.849999999999988,45.6C-4.849999999999988,45.6,0.4000000000000119,40,0.4000000000000119,40C0.4000000000000119,40,-4.849999999999988,34.4,-4.849999999999988,34.4"
                    fill="none"
                    stroke="#000000"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2; stroke-linejoin: round; stroke-linecap: round;"
                    transform="rotate(180 1.0500100000000114 40)"
                  />
                  <path
                    d="M144,40C144,40,144,40,1.0500000000000114,40"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2;"
                  />
                  <path
                    d="M29.999999999999993,48L29.999999999999993,32"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); shape-rendering: crispEdges; stroke-width: 2;"
                  />
                  <path
                    d="M58.49999999999999,48L58.49999999999999,32"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); shape-rendering: crispEdges; stroke-width: 2;"
                  />
                  <path
                    d="M86.99999999999999,48L86.99999999999999,32"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); shape-rendering: crispEdges; stroke-width: 2;"
                  />
                  <path
                    d="M115.49999999999999,48L115.49999999999999,32"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); shape-rendering: crispEdges; stroke-width: 2;"
                  />
                  <path
                    d="M144,48L144,32"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); shape-rendering: crispEdges; stroke-width: 2;"
                  />
                  <path
                    d="M172.5,48L172.5,32"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); shape-rendering: crispEdges; stroke-width: 2;"
                  />
                  <path
                    d="M201,48L201,32"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); shape-rendering: crispEdges; stroke-width: 2;"
                  />
                  <path
                    d="M229.5,48L229.5,32"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); shape-rendering: crispEdges; stroke-width: 2;"
                  />
                  <path
                    d="M258,48L258,32"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); shape-rendering: crispEdges; stroke-width: 2;"
                  />
                  <path
                    d="M29.999999999999993,48L29.999999999999993,32"
                    fill="none"
                    stroke="#11accd"
                    stroke-width="3.5"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); shape-rendering: crispEdges; stroke-width: 3.5;"
                  />
                  <path
                    d="M258,48L258,32"
                    fill="none"
                    stroke="#11accd"
                    stroke-width="3.5"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); shape-rendering: crispEdges; stroke-width: 3.5;"
                  />
                </svg>
                <div
                  style="position: absolute; left: 0px; top: 0px;"
                >
                  <div
                    style="position: absolute; width: 24px; height: 24px; left: 17.999999999999993px; top: 28px; filter: drop-shadow(0px 0px 5px rgba(0, 0, 0, 0.5)); transform: translateX(0px) translateY(0px) translateZ(0);"
                  >
                    <svg
                      height="24"
                      style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                      version="1.1"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <desc
                        style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                      >
                        Created with Raphaël
                      </desc>
                      <defs
                        style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                      />
                      <ellipse
                        cx="NaN"
                        cy="NaN"
                        fill="#71b307"
                        r="NaN"
                        rx="7"
                        ry="7"
                        stroke="#ffffff"
                        stroke-width="3"
                        style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 3;"
                        transform=""
                      />
                    </svg>
                  </div>
                </div>
                <div
                  style="position: absolute; left: 0px; top: 0px; z-index: 2;"
                >
                  <div
                    data-interactive-kind-for-testing="movable-point"
                    style="position: absolute; width: 48px; height: 48px; left: 5.999999999999993px; top: 16px; transform: translateX(0px) translateY(0px) translateZ(0);"
                  >
                    <svg
                      height="48"
                      style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                      version="1.1"
                      width="48"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <desc
                        style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                      >
                        Created with Raphaël
                      </desc>
                      <defs
                        style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                      />
                      <ellipse
                        cx="24"
                        cy="24"
                        fill="#000000"
                        opacity="0"
                        rx="24"
                        ry="24"
                        stroke="#000"
                        style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); opacity: 0; cursor: move;"
                      />
                    </svg>
                  </div>
                </div>
                <span
                  class="graphie-label"
                  data-math-formula="-4"
                  style="position: absolute; padding: 7px; color: black; left: 29.999999999999993px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="-3"
                  style="position: absolute; padding: 7px; color: black; left: 58.49999999999999px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="-2"
                  style="position: absolute; padding: 7px; color: black; left: 86.99999999999999px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="-1"
                  style="position: absolute; padding: 7px; color: black; left: 115.49999999999999px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="0"
                  style="position: absolute; padding: 7px; color: black; left: 144px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="1"
                  style="position: absolute; padding: 7px; color: black; left: 172.5px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="2"
                  style="position: absolute; padding: 7px; color: black; left: 201px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="3"
                  style="position: absolute; padding: 7px; color: black; left: 229.5px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="4"
                  style="position: absolute; padding: 7px; color: black; left: 258px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="-4"
                  style="position: absolute; padding: 7px; color: rgb(17, 172, 205); left: 29.999999999999993px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="4"
                  style="position: absolute; padding: 7px; color: rgb(17, 172, 205); left: 258px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`number-line widget should snapshot: first render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph perseus-paragraph-centered"
      data-perseus-paragraph-index="0"
    >
      <div
        class="perseus-block-math"
      >
        <div
          class="perseus-block-math-inner"
          style="overflow-x: auto; overflow-y: hidden; padding-top: 10px; padding-bottom: 10px; margin-top: -10px; margin-bottom: -10px;"
        >
          <span
            class="mock-TeX"
          >
            E=2.5
          </span>
        </div>
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="1"
    >
      <div
        class="paragraph"
      >
        <strong>
          Move the dot to 
          <span
            style="white-space: nowrap;"
          >
            <span />
            <span
              class="mock-TeX"
            >
              -E
            </span>
            <span />
          </span>
           on the number line.
        </strong>
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="2"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <div
            class="perseus-widget perseus-widget-interactive-number-line"
          >
            <div
              class="graphie-container"
            >
              <div
                class="graphie"
                style="position: relative; width: 459.99999999999994px; height: 80px;"
              >
                <svg
                  height="80"
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); z-index: 2;"
                  version="1.1"
                  width="460"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <desc
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                  >
                    Created with Raphaël
                  </desc>
                  <defs
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                  />
                  <rect
                    fill="#000000"
                    height="80"
                    opacity="0"
                    r="0"
                    rx="0"
                    ry="0"
                    stroke="#000"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); opacity: 0;"
                    width="460"
                    x="0"
                    y="0"
                  />
                </svg>
                <svg
                  height="80"
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                  version="1.1"
                  width="459.99999999999994"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <desc
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                  >
                    Created with Raphaël
                  </desc>
                  <defs
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                  />
                  <path
                    d="M454.44999999999993,45.6C454.79999999999995,43.5,458.6499999999999,40.35,459.69999999999993,40C458.6499999999999,39.65,454.79999999999995,36.5,454.44999999999993,34.4"
                    fill="none"
                    stroke="#000000"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2; stroke-linejoin: round; stroke-linecap: round;"
                    transform=""
                  />
                  <path
                    d="M229.99999999999997,40C229.99999999999997,40,229.99999999999997,40,458.94999999999993,40"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2;"
                  />
                  <path
                    d="M-3.4499999999999886,45.6C-3.0999999999999885,43.5,0.7500000000000115,40.35,1.8000000000000114,40C0.7500000000000113,39.65,-3.099999999999989,36.5,-3.4499999999999886,34.4"
                    fill="none"
                    stroke="#000000"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2; stroke-linejoin: round; stroke-linecap: round;"
                    transform="rotate(180 1.8000000000000114 40)"
                  />
                  <path
                    d="M229.99999999999997,40C229.99999999999997,40,229.99999999999997,40,1.0500000000000114,40"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2;"
                  />
                  <path
                    d="M29.999999999999982,48L29.999999999999982,32"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2;"
                  />
                  <path
                    d="M79.99999999999999,48L79.99999999999999,32"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2;"
                  />
                  <path
                    d="M129.99999999999997,48L129.99999999999997,32"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2;"
                  />
                  <path
                    d="M179.99999999999997,48L179.99999999999997,32"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2;"
                  />
                  <path
                    d="M229.99999999999997,48L229.99999999999997,32"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2;"
                  />
                  <path
                    d="M280,48L280,32"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2;"
                  />
                  <path
                    d="M330,48L330,32"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2;"
                  />
                  <path
                    d="M380,48L380,32"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2;"
                  />
                  <path
                    d="M430,48L430,32"
                    fill="none"
                    stroke="#000000"
                    stroke-width="2"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2;"
                  />
                  <path
                    d="M29.999999999999982,48L29.999999999999982,32"
                    fill="none"
                    stroke="#6495ed"
                    stroke-width="3.5"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 3.5;"
                  />
                  <path
                    d="M430,48L430,32"
                    fill="none"
                    stroke="#6495ed"
                    stroke-width="3.5"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 3.5;"
                  />
                </svg>
                <div
                  style="position: absolute; left: 0px; top: 0px;"
                >
                  <div
                    style="position: absolute; width: 34px; height: 34px; left: 12.999999999999982px; top: 23px; transform: translateX(0px) translateY(0px) translateZ(0);"
                  >
                    <svg
                      height="34"
                      style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                      version="1.1"
                      width="34"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <desc
                        style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                      >
                        Created with Raphaël
                      </desc>
                      <defs
                        style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                      />
                      <ellipse
                        cx="NaN"
                        cy="NaN"
                        fill="#71b307"
                        r="NaN"
                        rx="6"
                        ry="6"
                        stroke="#71b307"
                        stroke-width="1"
                        style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 1;"
                        transform=""
                      />
                    </svg>
                  </div>
                </div>
                <div
                  style="position: absolute; left: 0px; top: 0px; z-index: 2;"
                >
                  <div
                    data-interactive-kind-for-testing="movable-point"
                    style="position: absolute; width: 48px; height: 48px; left: 5.999999999999982px; top: 16px; transform: translateX(0px) translateY(0px) translateZ(0);"
                  >
                    <svg
                      height="48"
                      style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                      version="1.1"
                      width="48"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <desc
                        style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                      >
                        Created with Raphaël
                      </desc>
                      <defs
                        style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                      />
                      <ellipse
                        cx="24"
                        cy="24"
                        fill="#000000"
                        opacity="0"
                        rx="24"
                        ry="24"
                        stroke="#000"
                        style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); opacity: 0; cursor: move;"
                      />
                    </svg>
                  </div>
                </div>
                <span
                  class="graphie-label"
                  data-math-formula="-4"
                  style="position: absolute; padding: 7px; color: black; left: 29.999999999999982px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="-3"
                  style="position: absolute; padding: 7px; color: black; left: 79.99999999999999px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="-2"
                  style="position: absolute; padding: 7px; color: black; left: 129.99999999999997px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="-1"
                  style="position: absolute; padding: 7px; color: black; left: 179.99999999999997px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="0"
                  style="position: absolute; padding: 7px; color: black; left: 229.99999999999997px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="1"
                  style="position: absolute; padding: 7px; color: black; left: 280px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="2"
                  style="position: absolute; padding: 7px; color: black; left: 330px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="3"
                  style="position: absolute; padding: 7px; color: black; left: 380px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="4"
                  style="position: absolute; padding: 7px; color: black; left: 430px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="-4"
                  style="position: absolute; padding: 7px; color: black; left: 29.999999999999982px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
                <span
                  class="graphie-label"
                  data-math-formula="4"
                  style="position: absolute; padding: 7px; color: black; left: 430px; top: 61.2px; fill: none;"
                >
                  <span
                    class="tex-holder"
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
